extensions/positron-python/src/client/positron/session.ts

```diff
@@ -527,66 +522,62 @@ export class PythonRuntimeSession implements positron.LanguageRuntimeSession, vs
+     */
     async activateLsp(): Promise<void> {
-        // Start LSP for the foreground session only if its been previously stopped
-        if (this._lsp?.state === LspState.stopped) {
-            return this._queue.add(async () => {
-                await this.startLsp();
-            });
-        } else {
-            return undefined;
-        }
-    }
+        return this._queue.add(async () => {
+            if (!this._lsp) {
+                this._kernel?.emitJupyterLog('Tried to start LSP but no LSP instance available');
+                return;
+            }
 
-    async deactivateLsp(awaitStop: boolean): Promise<void> {
-        if (this._lsp?.state === LspState.running) {
-            return this._queue.add(async () => {
-                if (this._kernel) {
-                    this._kernel.emitJupyterLog(`Stopping Positron LSP server`);
-                }
-                await this._lsp?.deactivate(awaitStop);
-            });
-        } else {
-            return undefined;
-        }
-    }
+            if (this._lsp.state !== LspState.stopped && this._lsp.state !== LspState.uninitialized) {
+                // LSP is already started.
+                return;
+            }
+
+            if (!this._kernel) {
+                traceWarn('Cannot start LSP; kernel not started');
+                return;
+            }
 
-    private async startLsp(): Promise<void> {
-        if (this._kernel) {
             this._kernel.emitJupyterLog('Starting Positron LSP server');
 
             // Create the LSP comm, which also starts the LSP server.
             // We await the server selected port (the server selects the
             // port since it is in charge of binding to it, which avoids
             // race conditions). We also use this promise to avoid restarting
             // in the middle of initialization.
-            this._lspStarting = this._kernel.startPositronLsp('127.0.0.1');
-            const port = await this._lspStarting;
+            const port = await this._kernel.startPositronLsp('127.0.0.1');
 
             this._kernel.emitJupyterLog(`Starting Positron LSP client on port ${port}`);
 
-            await this._lsp?.activate(port);
-        }
+            await this._lsp.activate(port);
+        });
+    }
+
+    /**
+     * Stops the LSP if it is running
+     *
+     * Returns a promise that resolves when the LSP has been deactivated.
+     */
+    async deactivateLsp(): Promise<void> {
+        return this._queue.add(async () => {
+            if (!this._lsp || this._lsp.state !== LspState.running) {
+                // LSP is already stopped.
+                return;
+            }
+
+            this._kernel?.emitJupyterLog(`Stopping Positron LSP server`);
+            await this._lsp.deactivate(true);
+        });
     }
 
     async restart(workingDirectory?: string): Promise<void> {
         if (this._kernel) {
-            // Stop the LSP client before restarting the kernel. Don't stop it
```

Comment:
LSP management doesn't live here anymore, but this problem should be solved by the queue.